src/main/java/org/apache/xml/security/utils/KeyUtils.java

```diff
@@ -0,0 +1,280 @@
+     * @param provider provider to use for key generation
+     * @return generated keypair
+     * @throws XMLEncryptionException
+     */
+    public static KeyPair generateEphemeralDHKeyPair(PublicKey recipientPublicKey, Provider provider) throws XMLEncryptionException {
+        String algorithm = recipientPublicKey.getAlgorithm();
+        KeyPairGenerator keyPairGenerator;
+        try {
+
+            if (recipientPublicKey instanceof ECPublicKey) {
+                keyPairGenerator = createKeyPairGenerator(algorithm, provider);
+                ECPublicKey exchangePublicKey = (ECPublicKey) recipientPublicKey;
+                String keyOId = ECDSAUtils.getOIDFromPublicKey(exchangePublicKey);
+                if (keyOId == null) {
+                    keyOId = DERDecoderUtils.getAlgorithmIdFromPublicKey(recipientPublicKey);
+                }
+                ECGenParameterSpec kpgparams = new ECGenParameterSpec(keyOId);
+                keyPairGenerator.initialize(kpgparams);
+            } else {
+                String keyOId = DERDecoderUtils.getAlgorithmIdFromPublicKey(recipientPublicKey);
+                KeyType keyType = KeyType.getByOid(keyOId);
+                keyPairGenerator =  createKeyPairGenerator(keyType==null?keyOId: keyType.getName() , provider);
+            }
+
+            return keyPairGenerator.generateKeyPair();
+        } catch (NoSuchAlgorithmException | InvalidAlgorithmParameterException | DERDecodingException e) {
+            throw new XMLEncryptionException(e);
+        }
+    }
+
+    /**
+     * Create a KeyPairGenerator for the given algorithm and provider.
+     *
+     * @param algorithm  the key JCE algorithm name
+     * @param provider the provider to use or null if default JCE provider should be used
+     * @return the KeyPairGenerator
+     * @throws NoSuchAlgorithmException
+     */
+    public static KeyPairGenerator createKeyPairGenerator(String algorithm, Provider provider) throws NoSuchAlgorithmException {
+        return provider == null ? KeyPairGenerator.getInstance(algorithm)
+                : KeyPairGenerator.getInstance(algorithm, provider);
+    }
+
+
+    /**
+     * Method generates a secret key for given KeyAgreementParameterSpec.
+     *
+     * @param parameterSpec KeyAgreementParameterSpec which defines algorithm to derive key
+     * @return generated secret key
+     * @throws XMLEncryptionException
+     */
+    public static SecretKey aesWrapKeyWithDHGeneratedKey(KeyAgreementParameterSpec parameterSpec)
+            throws XMLEncryptionException {
+        try {
+            PublicKey publicKey = parameterSpec.getAgreementPublicKey();
+            PrivateKey privateKey = parameterSpec.getAgreementPrivateKey();
+
+            String algorithm = publicKey.getAlgorithm();
+            // java 11 uses ECDH instead of EC
+            algorithm = algorithm.equalsIgnoreCase("EC") ? "ECDH" : algorithm;
+            KeyAgreement keyAgreement = KeyAgreement.getInstance(algorithm);
```

Comment:
Hi @coheigea 
There is no other "KeyAgreement option for EC keys than ECDH. And I would like to keep them To test implementation against W3C test cases. But of course, the implementation also works with X25519 and X448, which is our final goal. Anyhow, I now put the warning in case of usage of the EC keys.